audio/out/buffer.c

```diff
@@ -425,10 +425,21 @@ void ao_set_paused(struct ao *ao, bool paused, bool eof)
 
     mp_mutex_unlock(&p->lock);
 
-    if (do_reset)
-        ao->driver->reset(ao);
-    if (do_start)
-        ao->driver->start(ao);
+    if (ao->driver->set_pause) {
+        if (do_reset) {
+            ao->driver->set_pause(ao, true);
+            p->end_time_ns -= mp_time_ns();
```

Comment:
I have to save `end_time_ns - mp_time_ns()` during pausing and restore `end_time_ns + mp_time_ns()` during resuming. I just reuse the `end_time_ns` field. Given it is not accessed during playback is paused, I think it is fine.

The reason why I need to save this relative value is that this relative value reflects the samples that have been queued to device. When the ao is paused, duration of `end_time_ns - mp_time_ns()` is already queued. Because we do not flush device buffer when using `set_paused`, this duration should be restored when the ao is resumed.

Reply:
I see, that makes sense.

IMO this should really use a separate variable. Being clever about lifetimes of values can only cause problems later.

Reply:
Agreed. Please also add comments about the intended purpose as you described here.

Reply:
My opinion is that `end_time_ns` goes into a de facto invalid state when playback pauses. It is an absolute timestamp. As the clock still goes during the paused period, the value of `end_time_ns` literately becomes invalid.

The only usage of `end_time_ns` is in `ao_get_delay`. It calculates the driver delay using `max(0, p->end_time_ns - mp_time_ns())`. This is correct during playing but wrong during paused as `end_time_ns` is fixed but `mp_time_ns()` continues to increase. In fact, I'd rather set `end_time_ns` to zero to avoid misuse of it when the playback is paused. Given `end_time_ns - mp_time_ns()` should be smaller than any valid timestamp, I think setting `end_time_ns` to `end_time_ns - mp_time_ns()` actually serve as sanity check to prevent `max(0, p->end_time_ns - mp_time_ns())` from result in any values other than zero during paused.

Another point is that `end_time_ns` is an private field so there won't be accesses of it from outside of `buffer.c`

Reply:
> In fact, I'd rather set `end_time_ns` to zero to avoid misuse of it when the playback is paused.

That sounds sensible.

> Given `end_time_ns - mp_time_ns()` should be smaller than any valid timestamp,

mpv's monotonic timestamps start at zero, so such a value could possibly be confused for a valid value.
But you're correct that it can never be a valid timestamp in the sense of the audio timing code (because it is in the past).

> Another point is that end_time_ns is an private field so there won't be accesses of it from outside of buffer.c

I know. There's enough potential for bugs in this very file :)